merkletree/nodekv.go

```diff
@@ -0,0 +1,132 @@
+package merkletree
+
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/coniks-sys/coniks-go/crypto"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/storage/kv"
+	"github.com/coniks-sys/coniks-go/utils"
+)
+
+var (
+	ErrorBadNodeIdentifier = errors.New("[merkletree] Bad node identifier")
+)
+
+func serializeKVKey(epoch uint64, prefixBits []bool) []byte {
+	// NodeKeyIdentifier + epoch + len(prefixBits) + index
+	index := util.ToBytes(prefixBits)
+	key := make([]byte, 0, 1+8+4+len(index))
+	key = append(key, NodeKeyIdentifier)
+	key = append(key, util.ULongToBytes(epoch)...)
+	key = append(key, util.IntToBytes(len(prefixBits))...)
+	key = append(key, index...)
+	return key
+}
+
+// storeToKV stores an interiorNode into db as following scheme:
+// [identifier, level, leftHash, rightHash]
+func (n *interiorNode) storeToKV(epoch uint64, prefixBits []bool, wb kv.Batch) {
+	buf := make([]byte, 0, 1+4+crypto.HashSizeByte*2)
+	buf = append(buf, InteriorNodeIdentifier)
+	buf = append(buf, util.IntToBytes(n.level)...)
+	buf = append(buf, n.leftHash...)
+	buf = append(buf, n.rightHash...)
+
+	wb.Put(serializeKVKey(epoch, prefixBits), buf)
+	n.leftChild.storeToKV(epoch, append(prefixBits, false), wb)
+	n.rightChild.storeToKV(epoch, append(prefixBits, true), wb)
+}
+
+// storeToKV stores a userLeafNode into db as following scheme:
+// [identifier, level, key+len, value+len, salt, index, commitment]
+func (n *userLeafNode) storeToKV(epoch uint64, prefixBits []bool, wb kv.Batch) {
+	buf := make([]byte, 0, 1+4+crypto.HashSizeByte*2+vrf.Size+len(n.key)+len(n.value)+4+4)
+	buf = append(buf, LeafIdentifier)
+	buf = append(buf, util.IntToBytes(n.level)...)
+	buf = append(buf, util.IntToBytes(len(n.key))...)
+	buf = append(buf, []byte(n.key)...)
+	buf = append(buf, util.IntToBytes(len(n.value))...)
+	buf = append(buf, n.value...)
+	buf = append(buf, n.salt...)
+	buf = append(buf, n.index...)
+	buf = append(buf, n.commitment...)
+
+	wb.Put(serializeKVKey(epoch, prefixBits), buf)
+}
+
+// storeToKV stores an emptyNode into db as following scheme:
+// [identifier, level, index]
+func (n *emptyNode) storeToKV(epoch uint64, prefixBits []bool, wb kv.Batch) {
+	buf := make([]byte, 0, 1+4+len(n.index))
+	buf = append(buf, EmptyBranchIdentifier)
+	buf = append(buf, util.IntToBytes(n.level)...)
+	buf = append(buf, n.index...)
+
+	wb.Put(serializeKVKey(epoch, prefixBits), buf)
+}
+
+func loadNode(db kv.DB, epoch uint64, prefixBits []bool) (MerkleNode, error) {
+	nodeBytes, err := db.Get(serializeKVKey(epoch, prefixBits))
+	if err == db.ErrNotFound() {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	return deserializeNode(nodeBytes), nil
+}
+
+func deserializeNode(buf []byte) MerkleNode {
+	switch buf[0] {
+	case InteriorNodeIdentifier:
```

Comment:
This is a modularity comment: Might be worth moving each of these cases into a helper function to clean this function up.